### ISSUE TITLE
setsockopt needs to happen before bind or connect

### ISSUE THREAD

**Issue by deepfryed** [andrewvc/em-zeromq#13]
If i need to set hwm, i need to do it before bind or connect happens. The Context#bind or Context#connect methods make this impossible at the moment.

http://api.zeromq.org/2-1:zmq-setsockopt

```
int zmq_setsockopt (void *socket, int option_name, const void *option_value, size_t option_len);
```

**Caution:** All options, with the exception of ZMQ_SUBSCRIBE, ZMQ_UNSUBSCRIBE and ZMQ_LINGER, only take effect for subsequent socket bind/connects.

**Comment by deepfryed** [comment]
had a quick stab at this myself, came up with https://github.com/deepfryed/em-zeromq  feel free to steal any relevant bits :)

**Comment by schmurfy** [comment]
why cloning the project instead of a fork ?

**Comment by deepfryed** [comment]
I wanted a clearer separation of context, socket and connection classes.

**Comment by schmurfy** [comment]
Yeah I understood that but if something does not work in the current gem it won't stay that way anyway...
So my real question is do you want to help improve the em-zeromq gem or do you want to create your own ?

In the later case please use another name.

**Comment by deepfryed** [comment]
i don't plan to release my code to rubygems, it's only for my own internal use. I initially hacked it to show you what I meant my mirroring the zeromq api but ending up doing the whole thing.

I don't mind helping to improve the em-zeromq gem but i think it needs some work.
1. I'd like it if you could break it down like I've done with context, socket & connection classes.
2. add decorators/delegators in context and connection classes to preserve the current api (while allowing for deprecation in a later version)

This will allow current users to keep using it without breaking their applications, while you'll have the freedom to refactor and fix the issues I have mentioned above.

**Comment by deepfryed** [comment]
I have to add, the code I wrote was just an exercise in learning for me. I'm not pushing you to make any changes, but I will gladly help to make any changes you see fit.

**Comment by schmurfy** [comment]
We never really got into cases where we needed to change hwm or other parameters but not having a socket surely posed some problems already (just look at Context#create) and I already asked myself if adding back a Socket class wouldn't be preferable.

I have not given enough thoughts but the current Connection class could well become the Socket class (with some changes of course).

I don't think breaking the api to be that bad considering that the gem is still in pre-1.0 state but we are completely hijacking this issue so I propose you open a new issue with the proposed change or even better a pull request (even incomplete) so you can start looking at what changes would be needed.
